### PR TITLE
[PA-1528] Update portworx methods to ensure compatibility with px-backup 2.5.2

### PR DESCRIPTION
**Pull request by kphalgun-px**
**What this PR does / why we need it**:
This PR updates portworx methods for compatibility with Px-Backup 2.5.2 and fixes issues in AfterSuite and CleanupCloudSettingsAndClusters.

**Which issue(s) this PR fixes** (optional)
Closes #PA-1528

**Special notes for your reviewer**:
Please review the following Jenkins builds:

~~v2.5.1~~

~~S3: https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/basic-system-test/job/px/job/s3/job/px-backup-system-test-vanilla/407/~~

~~Aetos Dashboard: https://aetos.pwx.purestorage.com/resultSet/testSetID/330024~~

~~NFS: https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/basic-system-test/job/px/job/NFS/job/PX-Cloud-Vanilla-NFS/132/~~

~~Aetos Dashboard: https://aetos.pwx.purestorage.com/resultSet/testSetID/330025~~

v 2.5.2

I have started a run on both the S3 and NFS nightly pipelines. I will provide an update once they are completed.

Thanks @ak-px for vendoring Px-Backup 2.5.2 API and @mkoppal-px for the comments on PR #1609.

### CODE REVIEW

File: drivers/backup/portworx/portworx.go [hunk 1]
```diff
@@ -221,47 +223,203 @@ func (p *portworx) CreateCloudCredential(ctx context.Context, req *api.CloudCred
...
+	if err != nil {
+		return nil, err
+	}
+	if reqWithClusterUID, ok := reqInterface.(*api.ClusterInspectRequest); ok {
+		return p.clusterManager.Inspect(ctx, reqWithClusterUID)
+	}
+	return nil, fmt.Errorf("expected *api.ClusterInspectRequest type after filling missing UID, but received %T", reqInterface)
 }
 
 func (p *portworx) EnumerateCluster(ctx context.Context, req *api.ClusterEnumerateRequest) (*api.ClusterEnumerateResponse, error) {
```
Comment (kphalgun-px) [inline drivers/backup/portworx/portworx.go:379]: I will add a detailed comment explaining why this had to be done towards the end of the review.

File: tests/common.go [hunk 2]
```diff
@@ -3512,6 +3538,20 @@ func DeleteBackup(backupName string, backupUID string, orgID string, ctx context
 	return backupDeleteResponse, err
 }
 
+func DeleteBackupWithClusterUID(backupName string, backupUID string, clusterUid string, orgID string, ctx context1.Context) (*api.BackupDeleteResponse, error) {
```
Comment (snigdha-px) [inline tests/common.go:3541]: Description missing

**Review by snigdha-px — APPROVED** [review]
Minor change. Rest everything looks good

**Comment by kphalgun-px** [comment]
> Minor change. Rest everything looks good Also, can we please see why the checks are failing

Ideally, the build should be based on the `px-backup/2.5.2-dev` branch, as I plan to merge my changes into it. However, the checks are currently running on the master branch, where px-backup-api 2.5.2 has not yet been vendored.

**Comment by ak-px** [comment]
@kphalgun-px , set the upstream branch to px-backup/2.5.2-dev and rebase user branch. This should solve the check failures.

From your branch-
git branch --set-upstream-to px-backup/2.5.2-dev <your-branch>
git pull --rebase

File: drivers/backup/portworx/portworx.go [hunk 3]
```diff
@@ -422,28 +679,145 @@ func (p *portworx) WaitForBackupLocationDeletion(
 	return nil
 }
 
+func (p *portworx) GetBackupLocationUID(ctx context.Context, orgID string, backupLocationName string) (string, error) {
+	backupLocationEnumerateReq := &api.BackupLocationEnumerateRequest{
+		OrgId: orgID,
+	}
+	enumerateRsp, err := p.EnumerateBackupLocation(ctx, backupLocationEnumerateReq)
```
[the remark this answers is not in the document]
Comment (kphalgun-px) [reply drivers/backup/portworx/portworx.go:686]: As you noted, using `EnumerateBackupLocation` instead of `EnumerateBackupLocationByUser` can cause issues when multiple backup locations have the same name. This is also true for other enumeration methods. Consider a case in which the user is deleted and an admin wants to remove that backup location created by the user, using `EnumerateBackupLocationByUser` would not work because the admin did not create it, making it impossible to get the UID and delete the backup location. I am still unsure how to handle duplicate names unless we pass the owner ID or the actual owner context.

[the remark this answers is not in the document]
Comment (snigdha-px) [reply drivers/backup/portworx/portworx.go:686]: @mkoppal-px 
Can you please suggest?
It is a corner case and we do not have any test cases with same name backup location for two users.
We can add a note so that we do not run into any issue in future?

**Comment by kphalgun-px** [comment]
> @kphalgun-px , set the upstream branch to px-backup/2.5.2-dev and rebase user branch. This should solve the check failures.
> 
> From your branch- git branch --set-upstream-to px-backup/2.5.2-dev your-branch git pull --rebase

Thanks. I do not want to rebase it yet to make it easier to review. I will do so towards the end of the review.

**Comment by ak-px** [comment]
@kphalgun-px , Please update runs with 2.5.2-dev image.

File: tests/common.go [hunk 4]
```diff
@@ -3591,6 +3662,24 @@ func DeleteSchedule(backupScheduleName string, clusterName string, orgID string,
...
+	backupDriver := Inst().Backup
+	bkpScheduleDeleteRequest := &api.BackupScheduleDeleteRequest{
+		OrgId: orgID,
+		Name:  backupScheduleName,
+		// DeleteBackups indicates whether the cloud backup files need to
+		// be deleted or retained.
+		DeleteBackups: true,
+		Uid:           backupScheduleUid,
+	}
+	_, err := backupDriver.DeleteBackupSchedule(ctx, bkpScheduleDeleteRequest)
```
[the remark this answers is not in the document]
Comment (kphalgun-px) [reply tests/common.go:3676]: If you want to delete the schedule as a user and then wait for its deletion as an admin, DeleteScheduleWithUID alone would not work. We would need both DeleteScheduleWithUID and DeleteScheduleWithUIDAndWait. But since we did not implement a DeleteScheduleAndWait, I did not add DeleteScheduleWithUIDAndWait.

[the remark this answers is not in the document]
Comment (ak-px) [reply tests/common.go:3676]: That's not a general use case to delete schedule from one user and wait from admin. The current DeleteSchedule has wait within, you can have the same implemented here.

**Comment by kphalgun-px** [comment]
> @kphalgun-px , Please update runs with 2.5.2-dev image.

Both the runs have failed due to intermittent issues. I have rerun them again.

File: tests/common.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kphalgun-px) [reply tests/common.go:3676]: Implemented DeleteScheduleWithUIDAndWait. Thanks.

File: tests/backup/backup_helper.go [hunk 5]
```diff
@@ -4171,3 +4159,105 @@ func CreateRuleForBackupWithMultipleApplications(orgID string, appList []string,
...
+func GetAllBackupScheduleNamesByOwnerID(ownerID string, orgID string, ctx context.Context) ([]string, error) {
+	isAdminCtx, err := portworx.IsAdminCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+	backupScheduleEnumerateReq := &api.BackupScheduleEnumerateRequest{
+		OrgId: orgID,
+		EnumerateOptions: &api.EnumerateOptions{
+			Owners: func() []string {
+				if isAdminCtx {
```
Comment (ak-px) [inline tests/backup/backup_helper.go:4207]: Schedule enumerate using ownerID is not supported. same as PB-4199

File: tests/backup/backup_helper.go [hunk 6]
```diff
@@ -4171,3 +4159,105 @@ func CreateRuleForBackupWithMultipleApplications(orgID string, appList []string,
...
+	}
+	backupEnumerateResp, err := Inst().Backup.EnumerateBackup(ctx, backupEnumerateReq)
+	if err != nil {
+		return nil, err
+	}
+	backupNames := make([]string, 0)
+	for _, backupObj := range backupEnumerateResp.GetBackups() {
+		if isAdminCtx {
+			backupNames = append(backupNames, backupObj.GetName())
+		} else {
```
Comment (ak-px) [inline tests/backup/backup_helper.go:4188]: why we need a else logic here, if its not admin context , There will be only respective user backup in enumerate resp.

Comment (kphalgun-px) [reply tests/backup/backup_helper.go:4188]: I am using these functions to validate both backup share and cluster backup share. If User 1 shares a backup with User 2, I can use the GetAllBackupNamesByOwnerID function, passing in owner ID of user 1 and user 2 (non-admin) ctx, to list the backup names and validate.

Comment (ak-px) [reply tests/backup/backup_helper.go:4188]: Make sense, Thanks for clarification.

File: tests/backup/backup_helper.go (hunk 5 above)
Comment (ak-px) [reply tests/backup/backup_helper.go:4207]: ideally it should  be disabled for schedule as well, can you once make sure while filtering with ownerID any other user schedule is not getting listed.

**Review by sn-px — APPROVED** [review]

**Comment by kphalgun-px** [comment]
Thank you for the reviews. These changes are now included in PR #1663.
